internal/oidc/provider/upstreamprovider/upsteam_provider.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020-2023 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package upstreamprovider
+
+import (
+	"context"
+	"net/url"
+
+	"golang.org/x/oauth2"
+	"k8s.io/apimachinery/pkg/types"
+
+	"go.pinniped.dev/internal/authenticators"
+	"go.pinniped.dev/pkg/oidcclient/nonce"
+	"go.pinniped.dev/pkg/oidcclient/oidctypes"
+	"go.pinniped.dev/pkg/oidcclient/pkce"
+)
+
+type RevocableTokenType string
+
+// These strings correspond to the token types defined by https://datatracker.ietf.org/doc/html/rfc7009#section-2.1
+const (
+	RefreshTokenType RevocableTokenType = "refresh_token"
+	AccessTokenType  RevocableTokenType = "access_token"
+)
+
+// RefreshAttributes contains information about the user from the original login request
+// and previous refreshes.
+type RefreshAttributes struct {
+	Username             string
+	Subject              string
+	DN                   string
+	Groups               []string
+	AdditionalAttributes map[string]string
+	GrantedScopes        []string
+}
+
+type UpstreamOIDCIdentityProviderI interface {
+	// GetName returns a name for this upstream provider. The controller watching the OIDCIdentityProviders will
+	// set this to be the Name of the CR from its metadata. Note that this is different from the DisplayName configured
+	// in each FederationDomain that uses this provider, so this name is for internal use only, not for interacting
+	// with clients. Clients should not expect to see this name or send this name.
+	GetName() string
+
+	// GetClientID returns the OAuth client ID registered with the upstream provider to be used in the authorization code flow.
+	GetClientID() string
+
+	// GetResourceUID returns the Kubernetes resource ID
+	GetResourceUID() types.UID
+
+	// GetAuthorizationURL returns the Authorization Endpoint fetched from discovery.
+	GetAuthorizationURL() *url.URL
+
+	// HasUserInfoURL returns whether there is a non-empty value for userinfo_endpoint fetched from discovery.
+	HasUserInfoURL() bool
+
+	// GetScopes returns the scopes to request in authorization (authcode or password grant) flow.
+	GetScopes() []string
+
+	// GetUsernameClaim returns the ID Token username claim name. May return empty string, in which case we
+	// will use some reasonable defaults.
+	GetUsernameClaim() string
+
+	// GetGroupsClaim returns the ID Token groups claim name. May return empty string, in which case we won't
+	// try to read groups from the upstream provider.
+	GetGroupsClaim() string
```

Comment:
Would there be value here in adding a helper that would return the groups as a `[]string{}`?

Reply:
Ah, no, I don't think so, given the link to `https://datatracker.ietf.org/doc/html/rfc7009#section-2.1` and what this struct represents.